src/array/cuda/gather_mm.cu

```diff
@@ -0,0 +1,334 @@
+                static_cast<DType*>(C->data),
+                static_cast<IdType*>(etype->data),
+                tot_num_rows,
+                k_start, k, n);
+        }
+    });
+}
+
+/* \brief Implementation of GatherMM operator where input matrix A is sorted
+ * according to relation types. Each relation type calls cuBLAS GEMM operator
+ * sequentially.
+ */
+template <int XPU, typename IdType, int bits>
+void gatherMM_SortedEtype(const NDArray A,
+              const NDArray B,
+              NDArray C,
+              const NDArray A_dim1_per_rel,
+              const NDArray B_dim1_per_rel,
+              bool a_trans, bool b_trans) {
+    SWITCH_BITS(bits, DType, {
+        auto device = runtime::DeviceAPI::Get(A->ctx);
+        assert(A_dim1_per_rel.NumElements() == B_dim1_per_rel.NumElements());
+        int64_t num_rel = A_dim1_per_rel.NumElements();
+        const DType *A_data = A.Ptr<DType>();
+        const DType *B_data = B.Ptr<DType>();
+        const IdType* A_rel_data = A_dim1_per_rel.Ptr<IdType>();
+        const IdType* B_rel_data = B_dim1_per_rel.Ptr<IdType>();
+        DType *C_data = C.Ptr<DType>();
+        int64_t A_offset = 0, B_offset = 0, C_offset = 0;
+        int64_t m, n, k;
+        DType alpha = 1., beta = 0.;
+
+        auto* thr_entry = runtime::CUDAThreadEntry::ThreadLocal();
+        if (!thr_entry->cublas_handle)
+            CUBLAS_CALL(cublasCreate(&(thr_entry->cublas_handle)));
+        CUBLAS_CALL(cublasSetStream(thr_entry->cublas_handle,
+            thr_entry->stream));
+
+        for (int etype = 0; etype < num_rel; ++etype) {
+            assert((a_trans) ? A_rel_data[etype] : A->shape[1] ==  \
+                (b_trans) ? B->shape[1] : B_rel_data[etype]);
+            m = A_rel_data[etype];  // rows of A
+            n = B->shape[1];  // cols of B
+            k = B_rel_data[etype];  // rows of B == cols of A
+
+            DType* A_trans_data = nullptr;
+            DType* B_trans_data = nullptr;
+            if (a_trans) {
+                A_trans_data = static_cast<DType*>(device->AllocWorkspace \
+                    (A->ctx, m * k * sizeof(DType)));
+                _Transpose(thr_entry->cublas_handle, A_data + A_offset, A_trans_data, m, k);
+            }
+            if (b_trans) {
+                B_trans_data = static_cast<DType*>(device->AllocWorkspace \
+                    (B->ctx, n * k * sizeof(DType)));
+                _Transpose(thr_entry->cublas_handle, B_data + B_offset, B_trans_data, k, n);
+            }
+            if (a_trans || b_trans) {
+                int64_t tmp = k;
+                if (a_trans)
+                    std::swap(m, k);
+                if (b_trans)  {
+                    k = tmp;
+                    std::swap(n, k);
+                }
+            }
+            int ldb = n, lda = k, ldc = n;
+            CUBLAS_CALL(cublasGemm<DType>(
+                thr_entry->cublas_handle,
+                CUBLAS_OP_N,
+                CUBLAS_OP_N,
```

Comment:
Are you seeing better performance manually transposing the matrices, and then calling cublas, as opposed to just setting the transpose flag in the cublas call?

Reply:
Actually, I didn't compare the performance.  Wanted to keep a generic cublas gemm call  with or without transposed versions. But you might be right. Transposing using flags might be faster.  Btw, we also are using cublas for the transpose. I think that's what you meant by "manual"?

Reply:
Yes, sorry, by "manually" I really meant explicitly transposing them with cublas. I'd expect better performance simply telling cublas the matrices are to be multiplied transposed in it's gemm call, as it can then do something like transposing only blocks in shared memory.

Reply:
You are right! By implicitly transposing, I got 10%~45% improvement. I experimented with feature length from 16 to 512 on a random hetero-graph. Will stick to the implicit transposition. Thanks.